### PR TITLE
Fix propeller crash when inferring literal type for an offloaded literal

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
## Why are the changes needed?

Followup to this ticket https://github.com/flyteorg/flyte/pull/5705
Fixes propeller crash when the type inferred is nil.

## What changes were proposed in this pull request?

The LiteralTypeFromLiteral function didn't handle Offloaded type

Along with this adds the following changes.
* Reading of offloaded literal in array node map task if the input itself is an offloaded literal. This is to support passing the inputs to all the sub nodes
* Promise input supports for offloaded literal
* Removes logging of the inputs outputs which might be large from datacatlog
* Adds hashing logic for offloaded literal which is used for caching
* adds flytectl support to o/p offloaded metadata uri
* bumps the maxDownloadMbs to 1000 MB to be inline with the max offloaded literal size

This does increase the memory profile for propeller whenever it needs to download large literal

## How was this patch tested?
Tested using sandbox and the changes working fine for the following examples . Tested for variation of caching aswell

```
import typing
from dataclasses import dataclass
from flytekit import task, workflow, map_task
from flytekit.types.file import FlyteFile
from mashumaro.mixins.json import DataClassJSONMixin

@dataclass
# class InnerDC(DataClassJSONMixin):
class InnerDC:
    a: bool
    b: int
    c: float
    d: list[int]
    e: dict[str, int]
    f: FlyteFile

@dataclass
# class DC(DataClassJSONMixin):
class DC:
    name: str
    other: str
    f: FlyteFile
    inner: InnerDC


@task(cache=True, cache_version="1.1")
def f(n: int) -> DC:
    with open("/tmp/abc", "w") as f:
        f.write("wow")
    name = "a" * 500_000
    return DC(
        name=name,
        other=name,
        f=FlyteFile("/tmp/abc"),
        inner=InnerDC(
            a=True,
            b=42,
            c=3.1415,
            d=[1,2,3],
            e={"a": 1, "b": 2},
            f=FlyteFile("/tmp/abc"),
        )
    )

@task(cache=True, cache_version="1.1")
def get_list(n: int) -> list[int]:
    return list(range(n))

@task(cache=True, cache_version="1.1")
def id_dataclass(dc: DC) -> DC:
    with open(dc.f, "r") as f:
        print(f.read())
    return dc


@task(cache=True, cache_version="1.1")
def consume_list_of_objects(xs: list[DC]) -> list[DC]:
    with open(xs[0].f, "r") as f:
        print(f.read())
    return xs[0:2]


@workflow
def working_wf(n: int = 15) -> list[DC]:
    xs=map_task(f)(n=get_list(n=n))
    return consume_list_of_objects(xs=xs)


@workflow
def passing_map_output_wf(n: int = 15) -> list[DC]:
    xs=map_task(f)(n=get_list(n=n))
    # Notice how we pass the offloaded literal to map_task
    return map_task(id_dataclass)(dc=xs) 

@task(cache=True, cache_version="1.1")
def consume_single_element(y: DC) -> DC:
    with open(y.f, "r") as f:
        print(f.read())
    return y

@workflow
def indexing_map_output_wf(n: int = 15):
    a=passing_map_output_wf(n=n)
    consume_single_element(y=a[0])
```

Notice `consume_list_objects` returns a slice since the outputting of offloaded data is yet to be supported in flytekit without which if the data is too large the datacatlog call errors out with grpc error due to max size breached

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]
i think this is fine, but cc @eapolinario

**Comment by wild-endeavor** [comment]
this doesn't break any flyte releases right?

**Comment by pmahindrakar-oss** [comment]
@wild-endeavor yes right now it doesn't since the feature is disabled, but this happens in peculiar case of map task accepting an offloaded literal. I am working on more elaborate fix for this as the current fix doesn't completely solve it.

**Comment by pmahindrakar-oss** [comment]
@wild-endeavor added more fixes for the crash along with array node support to read offloaded literal ,promise support etc,

File: flytepropeller/pkg/controller/nodes/array/handler.go [hunk 1]
```diff
@@ -192,11 +192,20 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 
 		size := -1
 		for _, variable := range literalMap.Literals {
+			if variable.GetOffloadedMetadata() != nil {
+				// variable will be overwritten with the contents of the offloaded data which contains the actual large literal.
+				// We need this for the map task to be able to create the subNodeSpec
```
Comment (katrogan) [inline flytepropeller/pkg/controller/nodes/array/handler.go:197]: just so I understand, this is also needed so the map task can index into the inputs as well too?

File: flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go [hunk 2]
```diff
@@ -122,7 +122,7 @@ func (m *CatalogClient) Get(ctx context.Context, key catalog.Key) (catalog.Entry
 		logger.Debugf(ctx, "DataCatalog failed to get artifact by tag %+v, err: %+v", tag, err)
 		return catalog.Entry{}, err
 	}
-	logger.Debugf(ctx, "Artifact found %v from tag %v", artifact, tag)
+	logger.Debugf(ctx, "Artifact found %v from tag %v", artifact.GetId(), tag)
```
Comment (katrogan) [inline flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go:125]: curious why these changes are necessary? do we ever send too big inputs/outputs to cache service? shouldn't we be using the offloaded literal?

File: flytepropeller/pkg/controller/nodes/common/utils.go [hunk 3]
```diff
@@ -79,6 +81,24 @@ func GetTargetEntity(ctx context.Context, nCtx interfaces.NodeExecutionContext)
...
 }
 
+// ReadLargeLiteral reads the offloaded large literal needed by array node task
+func ReadLargeLiteral(ctx context.Context, datastore *storage.DataStore,
+	tobeRead *idlcore.Literal) error {
+	if tobeRead.GetOffloadedMetadata() == nil {
+		return fmt.Errorf("unsupported type for reading offloaded literal")
+	}
+	// read the offloaded literal
+	dataReference := tobeRead.GetOffloadedMetadata().GetUri()
```
Comment (katrogan) [inline flytepropeller/pkg/controller/nodes/common/utils.go:91]: nit: worth checking that `tobeRead.GetOffloadedMetadata().GetUri()` is non-empty?

File: flytepropeller/pkg/controller/nodes/common/utils.go [hunk 4]
```diff
@@ -107,6 +127,12 @@ func OffloadLargeLiteral(ctx context.Context, datastore *storage.DataStore, data
 		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
 		return err
 	}
+	// compute the hash of the literal
+	literalDigest, err := pbhash.ComputeHash(ctx, toBeOffloaded)
```
Comment (katrogan) [inline flytepropeller/pkg/controller/nodes/common/utils.go:131]: nit: can we move this to the if block under

`if toBeOffloaded.GetHash() == "" {`

so we only compute it when necessary?

**Review by katrogan — APPROVED** [review]
LGTM but hopefully @hamersaw or @pvditt can review here too!

File: flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go:125]: Even they are not too big breaching the 10 MB limit, but anything less than that threshold will get logged which is also huge and hence removed it. Let me know if you disagree

File: flytepropeller/pkg/controller/nodes/common/utils.go (hunk 4 above)
Comment (pmahindrakar-oss) [reply flytepropeller/pkg/controller/nodes/common/utils.go:131]: I wanted to do something before overrwriting the value, moved the value assignment at the end along with this being computed in conditional

File: flytepropeller/pkg/controller/nodes/array/handler.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply flytepropeller/pkg/controller/nodes/array/handler.go:197]: yeah subNodeSpec covers that as it contains the inputs for each subnode

File: flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go (hunk 2 above)
Comment (katrogan) [reply flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go:125]: sounds good, I could still see this being useful, maybe we only log a deterministic prefix?

Comment (pmahindrakar-oss) [reply flytepropeller/pkg/controller/nodes/catalog/datacatalog/datacatalog.go:125]: It was logging an entire object and hence i think logging an identifier should good for long term too and we have api's to fetch the aritifact using the ID if we want to debug further. Logging a prefix of object converted to string format doesn't seem useful IMO

**Review by hamersaw — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
